File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/ShoppingListService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.entity.Item;
import at.ac.tuwien.sepm.groupphase.backend.entity.ItemStorage;
import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.ShoppingListCreationDto;
import at.ac.tuwien.sepm.groupphase.backend.entity.ShoppingList;
import at.ac.tuwien.sepm.groupphase.backend.exception.ValidationException;
import at.ac.tuwien.sepm.groupphase.backend.exception.NotFoundException;
import at.ac.tuwien.sepm.groupphase.backend.exception.ServiceException;
//import org.apache.tomcat.util.net.openssl.ciphers.Authentication;
import org.springframework.security.core.Authentication;

import java.util.List;

public interface ShoppingListService {

    /**
     * Creates a new Shoppinglist.
     *
     * @return DTO of newly created shoppingList
     */
    Long createNewShoppingList(ShoppingListCreationDto dto);

    /**
     * creates new shoppinglist if not existing.
     *
     * @return returns id of the shopping list
     */
    Long createNewShoppingList();

    /**
     * Gets the shopping list by id.
     *
     * @param id Id of shopping list to get
     * @return shopping list by id
     */
    ShoppingList getShoppingListByid(Long id);

    /**
     * Plans a recipe.
     * Checks which of the required ingredients are not present
     * and puts them on the public shopping-list, with the amount that is missing from the storage.
     *
     * @param recipeId  id of recipe that user wants to plan
     * @param userName of user who sent the request
     * @param numberOfPeople number of people the Recipe is planned for
     *
     * @return a List of all the ingredients that were added to the ShoppingList
     *
     * @throws ValidationException if the recipe or values in User are invalid
     * @throws NotFoundException if the recipe or the items in storage can not be found
     * @throws ServiceException if UnitOfQuantity of an ingredient and the stored Item are not compatible
     */
    List<ItemStorage> planRecipe(Long recipeId, String userName, Integer numberOfPeople);

    /**
     * puts all ingredients of a Recipe on the Shoppinglist.
     * regardless of what`s in the Storage
     *
     * @param recipeId  id of recipe that user wants to cook
     * @param userName of user who sent the request
     * @param numberOfPeople number of people the Recipe is planned for
     *
     * @return a List of all the ingredients that were added to the ShoppingList
     *
     * @throws ValidationException if the recipe or values in User are invalid
     * @throws NotFoundException if the recipe or the items in storage can not be found
     * @throws ServiceException if UnitOfQuantity of an ingredient and the stored Item are not compatible
     */
    List<ItemStorage> putRecipeOnShoppingList(Long recipeId, String userName, Integer numberOfPeople);

    /**
     * Insert a storage item to the shopping list.
     *
     * @param groupId is used to check if blueprint for this item exists in this group.
     *
     * @return inserted object
     */
    ItemStorage saveItem(ItemStorage itemStorage, Long id, Long groupId);

    /**
     * Change amount of a storage item of the shopping list.
     *
     * @param itemStorage item with changed amount
     *
     * @return inserted object
     */
    ItemStorage changeAmountOfItem(ItemStorage itemStorage, Long shoppingListId);

    /**
     * Finds all storage items in a shopping list.
     *
     * @return returns list of storage items
     */
    List<ItemStorage> findAllByShoppingListId(Long storageId);

    /**
     * Checks if shoppinglist exists.
     *
     * @return returns id of the shopping list
     */
    Long findShoppingListById(Long id);

    /**
     * search itemStorage items with matching parameter.
     *
     * @param itemStorage itemStorage with parameters
     * @return result of the search
     */
    List<ItemStorage> searchItem(ItemStorage itemStorage);

    /**
     * Find all item entries.
     *
     * @return ordered list of all item entries
     */
    List<Item> findAllItems();

    /**
     * Work off shoppinglist. A list of items  will be transfered from shoppinglist to storage.
     *
     * @param authentication of the logged in user
     * @param boughtItems the items which have been bought and will be transfered into the storage
     * @return a list of items which were tranfered into the storage
     */
    List<ItemStorage> workOffShoppingList(Authentication authentication, List<ItemStorage> boughtItems);


    void deleteItemById(Long itemId, Long shoppingListId);


}

File: frontend/src/app/services/shopping-list.service.ts
import {Injectable} from '@angular/core';
import {HttpBackend, HttpClient, HttpParams} from '@angular/common/http';
import {Globals} from '../global/globals';
import {Observable} from 'rxjs';
import {ShoppingList} from '../dtos/shopping-list';
import {Item} from '../dtos/item';


@Injectable({
  providedIn: 'root'
})
export class ShoppingListService {

  private shoppingListBaseUri: string = this.globals.backendUri + '/shoppinglist';
  private nakedHttpClient: HttpClient;
  private authHttpClient: HttpClient;

  constructor(private globals: Globals, handler: HttpBackend, private httpClient: HttpClient) {
    this.nakedHttpClient = new HttpClient(handler);
    this.authHttpClient = httpClient;
  }

  getPrivateShoppingList(): Observable<ShoppingList> {
    console.log('service: get private shoppinglist');
    return this.httpClient.get<ShoppingList>(this.shoppingListBaseUri + '/private');
  }

  getPublicShoppingList(): Observable<ShoppingList> {
    console.log('service: get public shoppinglist');
    return this.httpClient.get<ShoppingList>(this.shoppingListBaseUri + '/public');
  }

  addToPrivateShoppingList(item: Item): Observable<Item> {
    console.log('service: add item to private shoppinglist with id: ', item.id);
    return this.httpClient.post<Item>(this.shoppingListBaseUri + '/private', item);
  }

  addToPublicShoppingList(item: Item): Observable<Item> {
    console.log('service: add item to public shoppinglist with id: ', item.id);
    return this.httpClient.post<Item>(this.shoppingListBaseUri + '/public', item);
  }

  changeAmountOfItemInPrivateShoppingList(item: Item): Observable<Item> {
    console.log('service: change amount of item on private shoppinglist with id: ', item.id);
    return this.httpClient.put<Item>(this.shoppingListBaseUri + '/private/' + item.id, item);
  }

  changeAmountOfItemInPublicShoppingList(item: Item): Observable<Item> {
    console.log('service: change amount of item on public shoppinglist with id: ', item.id);
    return this.httpClient.put<Item>(this.shoppingListBaseUri + '/public/' + item.id, item);
  }

  planRecipe(id: number, numOfPeople: number): Observable<Item[]> {
    console.log('service: plan recipe with id: ' + id + 'for number of people: ' + numOfPeople);
    //return this.httpClient.put<Item[]>(this.shoppingListBaseUri+'/?recipeId='+id+'&numberOfPeople='+numOfPeople, id);
    const httpParams = new HttpParams()
      .set('recipeId', id)
      .set('numberOfPeople', numOfPeople);
    return this.httpClient.put<Item[]>(this.shoppingListBaseUri, httpParams.toString(), {params: httpParams});
  }

  putRecipeOnShoppingList(id: number, numOfPeople: number): Observable<Item[]> {
    console.log('service: put all ingredients to shoppinglist of recipe with id: ' + id + 'for number of people: ' + numOfPeople);
    //return this.httpClient.put<Item[]>(
    //  this.shoppingListBaseUri+'/putAllIngredientsOfRecipe/?recipeId='+id+'&numberOfPeople='+numOfPeople, id);
    const httpParams = new HttpParams()
      .set('recipeId', id)
      .set('numberOfPeople', numOfPeople);
    return this.httpClient.put<Item[]>(
      this.shoppingListBaseUri + '/putAllIngredientsOfRecipe', httpParams.toString(), {params: httpParams});
  }

  /**
   * Loads items from the backend with specific parameters
   */
  searchItems(params: string): Observable<Item[]> {
    console.log('Search for items');
    return this.httpClient.get<Item[]>(
      this.shoppingListBaseUri + '/search' + params);
  }

  addItemToShoppingList(item: Item): Observable<Item> {
    console.log('service: add item to shoppinglist: ', item);
    console.log(this.shoppingListBaseUri + '/newItem');
    return this.httpClient.post<Item>(this.shoppingListBaseUri + '/newItem', item);
  }

  findAll(): Observable<Item[]> {
    console.log('service: load all items of shoppinglist');
    return this.httpClient.get<Item[]>(this.shoppingListBaseUri + '/availableItems');
  }

  findAllItems(): Observable<Item[]> {
    console.log('load items');
    return this.httpClient.get<Item[]>(this.shoppingListBaseUri + '/items');
  }

  workOffShoppingList(boughtItems: Item[], shoppinglistId: number): Observable<Item[]> {
    console.log('service: work off shopping-list: ' + boughtItems);
    return this.httpClient.put<Item[]>(this.shoppingListBaseUri + '/' + shoppinglistId, boughtItems);
  }

  getGroupStorageForUser(): Observable<number> {
    console.log('service: get group storage for user');
    return this.httpClient.get<number>(this.globals.backendUri + '/group/storage');
  }

  getGroupShoppingListForUser(): Observable<number> {
    console.log('service: get group shoppinglist for user');
    return this.httpClient.get<number>(this.globals.backendUri + '/group/shoppinglist');
  }

  deleteItemFromPrivate(id: number): Observable<object> {
    console.log('service: delete item of shoppinglist');
    return this.httpClient.delete<object>(this.shoppingListBaseUri + '/private/' + id);
  }

  deleteItemFromPublic(id: number): Observable<object> {
    console.log('service: delete item of shoppinglist');
    return this.httpClient.delete<object>(this.shoppingListBaseUri + '/public/' + id);
  }
}
